### PR TITLE
BUG: Add frequency to DatetimeArray/TimedeltaArray take 

### LINKED ISSUE #58327: BUG: get_level_values() in a MultiIndex returns DatetimeIndex without frequency even though the index level has a frequency (and unique values).
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> import pandas as pd
>>> dates  = pd.date_range("2000", freq = 'YS', periods = 4)
>>> index = ["A"]
>>> multi_index = pd.MultiIndex.from_product([dates, index], names = [“Dates”, “Second Index”])
>>> print(multi_index.levels[0].freq)
<YearBegin: month=1>
>>> print(multi_index.get_level_values(0).freq)
None
>>> print(multi_index.get_level_values(“Dates”).freq)
None
```

### Issue Description

For a DatetimeIndex contained in a MultiIndex get_level_values() returns a DatetimeIndex without frequency, even if the frequency is set and available in the internal structure (i.e. if accessed via index.levels, see example above).
 
The underlying issue seems to be the procedure in  _get_level_values(…) in index/multi.py :  When creating the object for the shallow copy a new DatetimeIndex is generated which does not carry along the DatetimeIndex frequency. 

Why (at least to me) it is important that this works correctly although it works with level[0].freq? With get_level_values() it is also possible to use the name as an identifier (not only the index level number), which in many cases is a lot more convenient. 

### Expected Behavior

I would expect that get_level_values() returns a DatetimeIndex with correct frequency, mirroring the one actually set in the MultiIndex.

### Installed Versions

<details>

------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.9.13.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19045
machine               : AMD64
processor             : Intel64 Family 6 Model 85 Stepping 0, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_United Kingdom.1252
 
pandas                : 2.2.2
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 69.2.0
pip                   : 24.0
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

<details>

### PR DESCRIPTION
**Pull request by annika-rudolph**
- [x] closes #58327 #57949
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

<details> 
<summary> Disclaimer </summary>
I am required to add that the software is provided "as is", without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and noninfringement. In no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software.
</details>

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -2358,6 +2355,35 @@ def interpolate(
         if not copy:
             return self
         return type(self)._simple_new(out_data, dtype=self.dtype)
+    
+    def take(
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:2359]: these classes already have a take method.  can you edit that?  or call super().take(...)?

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -2358,6 +2355,35 @@ def interpolate(
...
+            allow_fill=allow_fill,
+            fill_value=fill_value,
+            axis=axis,
+        )    
+        result = self._from_backing_data(new_data)
+        indices = np.asarray(indices, dtype=np.intp)
+        maybe_slice = lib.maybe_indices_to_slice(indices, len(self))
+        
+        if isinstance(maybe_slice, slice):
+            freq = self._get_getitem_freq(maybe_slice)
```
[the remark this answers is not in the document]
Comment (annika-rudolph) [reply pandas/core/arrays/datetimelike.py:2383]: A possibility could be to check for DatetimeArray/TimedeltaArray when creating the shallow copy of the level & add the frequency if applicable. The current approach seemed a bit more elegant to me (and judging from the tests it doesn't do any harm), but that's debatable for sure. I can check the solution that I mentioned.

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (annika-rudolph) [reply pandas/core/arrays/datetimelike.py:2359]: If you mean call super().take(...) and then extend it by the bit in the end to add the frequency: Yep, sure, that's more elegant :)

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -1244,7 +1244,9 @@ def _constructor(self) -> Callable[..., MultiIndex]:  # type: ignore[override]
         return type(self).from_tuples
 
     @doc(Index._shallow_copy)
-    def _shallow_copy(self, values: np.ndarray, name=lib.no_default) -> MultiIndex:
+    def _shallow_copy(
```
Comment (jbrockmendel) [inline pandas/core/indexes/multi.py:1247]: if the options are doing this in take or shallow_copy, take is definitely preferred

Comment (annika-rudolph) [reply pandas/core/indexes/multi.py]: done -- implemented via take

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -2369,6 +2370,27 @@ def interpolate(
...
+        result = super().take(
+            indices=indices, allow_fill=allow_fill, fill_value=fill_value, axis=axis
+        )
+
+        indices = np.asarray(indices, dtype=np.intp)
+        maybe_slice = lib.maybe_indices_to_slice(indices, len(self))
+
+        if isinstance(maybe_slice, slice):
+            freq = self._get_getitem_freq(maybe_slice)
+            result.freq = freq
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:2390]: i think if you set result._freq it will be more performant

Comment (annika-rudolph) [reply pandas/core/arrays/datetimelike.py]: changed

Comment (annika-rudolph) [reply pandas/core/arrays/datetimelike.py]: Hi @jbrockmendel is this now corresponding to your idea?

Comment (annika-rudolph) [reply pandas/core/arrays/datetimelike.py]: Hi @jbrockmendel , any news on this? would be nice to finish the PR at some point :)

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @annika-rudolph (sorry for the delay here)

**Pull request merged**